Sources/TransactionalProperty.swift

```diff
@@ -0,0 +1,690 @@
+
+	init<T: _TransactionalPropertyProtocol>(
+		_ inner: T,
+		transform: @escaping (T.Value) -> Value,
+		errorTransform: @escaping (T.TransactionError) -> TransactionError,
+		_ body: @escaping (Value) -> Result<T.Value, TransactionError>
+	)
+
+	init<T: _TransactionalPropertyProtocol>(
+		_ inner: T,
+		transform: @escaping (T.Value) -> Value,
+		_ body: @escaping (Value) -> Result<T.Value, TransactionError>
+	) where T.TransactionError == NoError
+}
+
+extension TransactionalProperty: _TransactionalPropertyProtocol {
+	public var property: TransactionalProperty<Value, TransactionError> {
+		return self
+	}
+}
+
+public enum ValidationResult<Value, Error: Swift.Error> {
+	case success(Value)
+	case failure(Value, Error)
+
+	public var isFailed: Bool {
+		if case .failure = self {
+			return true
+		} else {
+			return false
+		}
+	}
+
+	public var error: Error? {
+		if case let .failure(_, error) = self {
+			return error
+		} else {
+			return nil
+		}
+	}
+
+	public var result: Result<Value, Error> {
+		switch self {
+		case let .success(value):
+			return .success(value)
+		case let .failure(_, error):
+			return .failure(error)
+		}
+	}
+
+	fileprivate init<InnerValue>(_ innerValue: InnerValue, transform: (InnerValue) -> Value, validator: (Value) -> Result<InnerValue, Error>) {
+		let value = transform(innerValue)
+
+		switch validator(value) {
+		case .success:
+			self = .success(value)
+		case let .failure(error):
+			self = .failure(value, error)
+		}
+	}
+
+	public func map<U, E: Swift.Error>(_ transform: (Value) -> U, _ errorTransform: (Error) -> E, validator: (U) -> Result<Value, E>) -> ValidationResult<U, E> {
+		switch self {
+		case let .success(value):
+			switch validator(transform(value)) {
+			case let .success(innerValue):
+				return .success(transform(innerValue))
+			case let .failure(error):
+				return .failure(transform(value), error)
+			}
+
+		case let .failure(value, error):
+			return .failure(transform(value), errorTransform(error))
+		}
+	}
+}
+
+// FIXME: Remove the type parameter that works around type checker weirdness.
+private class TransactionalPropertyBox<Property: ComposableMutablePropertyProtocol>: TransactionalPropertyBoxBase<()> {
+	private let base: Property
+
+	init(_ base: Property) { self.base = base }
+
+	override func lock<R>(_ action: (()) throws -> R) rethrows -> R {
+		return try base.withValue { _ in return try action(()) }
+	}
+}
+
+private class TransactionalPropertyBoxBase<Value> {
+	func lock<R>(_ action: (()) throws -> R) rethrows -> R {
+		fatalError()
+	}
+}
+
+extension ComposableMutablePropertyProtocol {
+	// - The default `map` with `NoError`.
+
+	/// Create a mutable mapped view to `self`.
+	///
+	/// - parameters:
+	///   - forward: The value transform to convert `Self.Value` to `U`.
+	///   - backward: The value transform to convert `U` to `Self.Value`.
+	///
+	/// - returns: A mapping `TransactionalProperty`.
+	public func map<U>(
```

Comment:
If we changed the name from `TransactionalProperty` to `PropertyEditor`, and call the "value transformers" as they are, we could instead use this sort of signature:

```swift
public func editor<U>(
    transform: @escaping (Value) -> U,
    reverseTransform: @escaping (U) -> Value
) -> PropertyEditor<U, NoError> {
    return PropertyEditor(self, transform: transform) { proposedInput in
        return .success(reverseTransform(proposedInput))
    }
}
```

Or, if we use the `PropertyTransformer` naming, this could be:

```swift
public func valueTransformer<U>(
    forward: @escaping (Value) -> U,
    reverse: @escaping (U) -> Value
) -> PropertyEditor<U, NoError> {
    return PropertyTransformer(self, transform: forward) { proposedInput in
        return .success(reverse(proposedInput))
    }
}
```

I used `reverse` since it better matches the `NSValueTransformer` conventions. Also, something about `reverse` sounds more correct than `backward` to me for some reason.

Thoughts?